### DISCUSSION TITLE
Expose Native Objective-C OR Swift module created in react native bridge to the Application's native side

### DISCUSSION THREAD

**Discussion by rakesho-vel** [callstack/react-native-builder-bob#553] [Q&A] [answered]
I have created a new react native library and working on creating a native module for push notification. I want to get the push notification token in my react native bridge from application's AppDelegate.m.
For this, i am trying to create a new Objective-C object in react native bride's ios folder and trying to expose it to the App's AppDelegate.mm file, but that class/header is unable to find to the AppDelegate file.

Do i need to do anything extra to make this class/module from react-native bridge folder accessible to the react native application's Native code?

Adding screenshot for the reference - 
![Screenshot 2024-05-10 at 8 04 50 PM](https://github.com/callstack/react-native-builder-bob/assets/84490964/365af80d-216f-413f-9c18-3503f28f91a3)

**Comment by atlj** [comment]
Not sure if this helps, but here is another approach,

In your `AppDelegate`, import your library header. Simply `#import RnPushNotification.h` and call a method that sets the token. Something like `[RnPushNotification setToken:token];`. This should be more straightforward than trying to import the app delegate in the library. So its essentially the other way around.

Please check how [React Native Bootsplash](https://github.com/zoontek/react-native-bootsplash?tab=readme-ov-file#ios-1) handles it. They follow a similar approach.

**Comment by rakesho-vel** [comment]
@atlj yes, that is what I am trying to do, sorry, if i was misunderstood. Here is my detailed code and the error i am getting in XCode - 

RnPushNotification.h
````
#import <Foundation/Foundation.h>

NS_ASSUME_NONNULL_BEGIN

@interface RnPushNotification : NSObject

+ (void) setToken:token;

@end

NS_ASSUME_NONNULL_END

````

RnPushNotification.m

````
#import "RnPushNotification.h"

@implementation RnPushNotification

+ (void)setToken:(id)token {
//  implementation to set the token
}
@end

````
When I import this method in the ReactNative Example Application's AppDelegate.m file, i am getting File not found error
<img width="997" alt="Screenshot 2024-05-10 at 9 46 41 PM" src="https://github.com/callstack/react-native-builder-bob/assets/84490964/05e4aed0-d834-4636-95f2-c2681e517815">


PLEASE NOTE: Currently I am trying to add this function in the example app i got with the Scaffold command.

**Reply by atlj** [reply]
Quite interesting. When I try to do the same (import a header from a crnl generated project), it works just fine. I guess you might have some bad cache or something like that? Could you clear all your build files and derived data? You can use something like [Xclean](https://github.com/macmade/Xclean)

**Reply by rakesho-vel** [reply]
Yes, I tried clearing the build, and cleared the derived data, did restart everything as well but getting same error :( 
There is one default bridging header in the react native bridge folder and i can only import that but not the newly added objective-c file OR the default module i got in the react native bridge.

![Screenshot 2024-05-10 at 8 04 50 PM](https://github.com/callstack/react-native-builder-bob/assets/84490964/cbf101fa-b1fa-4ac6-9650-f2f9eabf1181)

**Reply by atlj** [reply]
Could you also send your `.podspec` file? I'm curious if those `.h` and `.m` files are getting included

**Reply by rakesho-vel** [reply]
Okay, here is my .podspec file - 

```
require "json"

package = JSON.parse(File.read(File.join(__dir__, "package.json")))
folly_compiler_flags = '-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1 -Wno-comma -Wno-shorten-64-to-32'

Pod::Spec.new do |s|
  s.name         = "react-native-push"
  s.version      = package["version"]
  s.summary      = package["description"]
  s.homepage     = package["homepage"]
  s.license      = package["license"]
  s.authors      = package["author"]

  s.platforms    = { :ios => min_ios_version_supported }
  s.source       = { :git => "https://github.com/rakesho-vel/react-native-push.git", :tag => "#{s.version}" }

  s.source_files = "ios/**/*.{h,m,mm,swift}"

  # Use install_modules_dependencies helper to install the dependencies if React Native version >=0.71.0.
  # See https://github.com/facebook/react-native/blob/febf6b7f33fdb4904669f99d795eba4c0f95d7bf/scripts/cocoapods/new_architecture.rb#L79.
  if respond_to?(:install_modules_dependencies, true)
    install_modules_dependencies(s)
  else
  s.dependency "React-Core"

  # Don't install the dependencies when we run `pod install` in the old architecture.
  if ENV['RCT_NEW_ARCH_ENABLED'] == '1' then
    s.compiler_flags = folly_compiler_flags + " -DRCT_NEW_ARCH_ENABLED=1"
    s.pod_target_xcconfig    = {
        "HEADER_SEARCH_PATHS" => "\"$(PODS_ROOT)/boost\"",
        "OTHER_CPLUSPLUSFLAGS" => "-DFOLLY_NO_CONFIG -DFOLLY_MOBILE=1 -DFOLLY_USE_LIBCPP=1",
        "CLANG_CXX_LANGUAGE_STANDARD" => "c++17"
    }
    s.dependency "React-Codegen"
    s.dependency "RCT-Folly"
    s.dependency "RCTRequired"
    s.dependency "RCTTypeSafety"
    s.dependency "ReactCommon/turbomodule/core"
   end
  end  
end

```

**Reply by rakesho-vel** [reply]
So i tried creating a package for this SDK and installed it in new example project i created, and surprisingly it is able to import there - 
<img width="1415" alt="Screenshot 2024-05-11 at 9 12 50 AM" src="https://github.com/callstack/react-native-builder-bob/assets/84490964/2c424de8-041d-4436-863f-d2086d66ce8d">

I also created another react native sdk project using scaffold command and added new Objective-C class there but it is also not able to import in its example application.

**Reply by atlj** [reply]
Interesting... I'm starting to suspect the Podfile here, is your Podfile different than the other project (the Podfile not podspec). Other than that you podspec looks ok. Maybe install bootsplash to your example app and see if you can import its header?

**Reply by rakesho-vel** [reply]
I haven't made any changes to Podfile (example app's - react-native-sdk/example/ios/Podfile), here is the podfile the react native builder generated - 

```
# Resolve react_native_pods.rb with node to allow for hoisting
require Pod::Executable.execute_command('node', ['-p',
  'require.resolve(
    "react-native/scripts/react_native_pods.rb",
    {paths: [process.argv[1]]},
  )', __dir__]).strip

platform :ios, min_ios_version_supported
prepare_react_native_project!

# If you are using a `react-native-flipper` your iOS build will fail when `NO_FLIPPER=1` is set.
# because `react-native-flipper` depends on (FlipperKit,...) that will be excluded
#
# To fix this you can also exclude `react-native-flipper` using a `react-native.config.js`
# ```js
# module.exports = {
#   dependencies: {
#     ...(process.env.NO_FLIPPER ? { 'react-native-flipper': { platforms: { ios: null } } } : {}),
# ```
flipper_config = ENV['NO_FLIPPER'] == "1" ? FlipperConfiguration.disabled : FlipperConfiguration.enabled

linkage = ENV['USE_FRAMEWORKS']
if linkage != nil
  Pod::UI.puts "Configuring Pod with #{linkage}ally linked Frameworks".green
  use_frameworks! :linkage => linkage.to_sym
end

target 'hydraexamplelatest' do
  config = use_native_modules!

  use_react_native!(
    :path => config[:reactNativePath],
    # Enables Flipper.
    #
    # Note that if you have use_frameworks! enabled, Flipper will not work and
    # you should disable the next line.
    :flipper_configuration => flipper_config,
    # An absolute path to your application root.
    :app_path => "#{Pod::Config.instance.installation_root}/.."
  )

  target 'hydraexamplelatestTests' do
    inherit! :complete
    # Pods for testing
  end

  post_install do |installer|
    # https://github.com/facebook/react-native/blob/main/packages/react-native/scripts/react_native_pods.rb#L197-L202
    react_native_post_install(
      installer,
      config[:reactNativePath],
      :mac_catalyst_enabled => false
    )
  end
end

```

FYII have generated sdk project with this command - 
```
npx create-react-native-library@latest react-native-awesome-module
```

**Reply by rakesho-vel** [reply, marked as answer]
Okay, this time (4th OR 5th time), i removed all node modules in example app, reinstalled node as well as pod packages and now it is able to import. Not sure what was the issue as it did not work last times. 

Thank you so much for the quick replies and help though 🙏

**Reply by atlj** [reply]
Nice! Glad it worked out for you!
